### PR TITLE
Scrollable frame

### PR DESCRIPTION
**Pull request by dorileo**
This PR includes a small fix, a theme change and a new feature for OnKeyDown events callback calls and more importantly a new feature for frame control. We can now make it scrollable.

@VladimirMarkelov can you please help me in depth reviewing the "frame: scrollable frame" patch? this patch touches very sensible areas and I faced many corner cases while implementing it, I suppose you may find a few more. Can you also test with you current test cases for frame?

Thanks...

### CODE REVIEW

**Comment by VladimirMarkelov** [comment]
A very good one. Thank you! It is long, so it takes some time for me to read and understand. Please be patient ;)

File: window.go [hunk 1]
```diff
@@ -21,8 +21,14 @@ type Window struct {
...
 
 	onClose        func(Event) bool
-	onKeyDown      func(Event) bool
 	onScreenResize func(Event)
+
+	onKeyDown *keyDownCb
+}
+
+type keyDownCb struct {
+	data interface{}
```
Comment (VladimirMarkelov) [inline window.go:30]: Question: how is `data` going to be used? In the patch I do not see an example that explains its usage.

File: theme.go [hunk 2]
```diff
@@ -510,9 +510,11 @@ func ThemeInfo(name string) ThemeDesc {
 // color from the theme.
 // clr - current object color
 // id - color ID in theme
-func RealColor(clr term.Attribute, id string) term.Attribute {
+func RealColor(clr term.Attribute, style string, id string) term.Attribute {
```
Comment (VladimirMarkelov) [inline theme.go]: Nitpick: I think function comment should be updated to explain new argument

File: theme.go [hunk 3]
```diff
@@ -510,9 +510,11 @@ func ThemeInfo(name string) ThemeDesc {
...
+	ccolor := fmt.Sprintf("%s%s", style, id)
```
Comment (VladimirMarkelov) [inline theme.go]: Just a question: Can you provide an example of `style`? I see they are just concatenated. Is a point between them make a resulting string better: `ccolor := fmt.Sprintf("%s.%s", style, id)`?

File: frame.go [hunk 4]
```diff
@@ -67,16 +79,52 @@ func (f *Frame) Draw() {
...
+				prop = int(math.Round((float64(h) / (100 / frameProp))))
+			}
+
+			f.lastScrollProp = prop
+		}
+
+		DrawScrollBar(x+w, y, 1, h, f.lastScrollProp)
+	}
+
+	fg, bg := RealColor(f.fg, f.Style(), ColorViewText), RealColor(f.bg, f.style, ColorViewBack)
```
Comment (VladimirMarkelov) [inline frame.go]: Why one uses `f.Style()` while another one `f.style`?

File: ctrlutil.go [hunk 5]
```diff
@@ -266,8 +266,59 @@ func SendEventToChild(parent Control, ev Event) bool {
...
+
+	return clipped
+}
+
+// ControlInRect returns true if c is within a given rect
+func ControlInRect(c Control, x int, y int, w int, h int) bool {
+	xx, yy := c.Pos()
+	ww, hh := c.Size()
+
+	return xx >= x && xx >= x &&
```
Comment (VladimirMarkelov) [inline ctrlutil.go]: `xx >= x && xx >= x` looks suspicious ;)

**Review by VladimirMarkelov — COMMENTED** [review]
I looked through the changes but have not tried them yet.

File: window.go (hunk 1 above)
Comment (dorileo) [reply window.go:30]: data is used by the callback user, i.e:

```go
type UserData struct {
...
}
myData := &UserData{}
wnd.wnd.OnKeyDown(func(evt Event, data interface{}) bool{
     md := data.(UserData)
     // do some data manipulation here this could be some key handling logic i.e alt+<key>
    // manipulation where the first hit detects alt then do some special handling after that
   // and so on
}, myData)

```

File: theme.go (hunk 2 above)
Comment (dorileo) [reply theme.go]: sure

File: theme.go (hunk 3 above)
Comment (dorileo) [reply theme.go]: sure, I can provide some examples. I like the suggestion of changing it to ```<style>.<id>```

File: frame.go (hunk 4 above)
Comment (dorileo) [reply frame.go]: No reason at all, just a careless and inconsistent programmer :)

File: ctrlutil.go (hunk 5 above)
Comment (dorileo) [reply ctrlutil.go]: another careless programmer speaking here :)

**Comment by VladimirMarkelov** [comment]
I tried the patch - applications work fine. So, I'll wait for your changes ad we can merge it ;)

**Comment by VladimirMarkelov** [comment]
And it would be great to have a small example of how to use scrolling frames.

**Comment by dorileo** [comment]
@VladimirMarkelov updated the set with the requested changes + a patch introducing a small demo with a scrollable frame.
